[docs][guides] Add keyboard handling guide (#30983)
expo/expo: docs/pages/guides/keyboard-handling.mdx

---
title: Keyboard handling
sidebar_title: Keyboard handling
description: A guide to handling common keyboard interactions.
---

Keyboard handling is crucial for creating an excellent user experience in your Expo app. This guide covers common keyboard interactions and how to manage them effectively.

> React Native provides several keyboard-related APIs, with [Keybaord](https://reactnative.dev/docs/keyboard) and [KeyboardAvoidingView](https://reactnative.dev/docs/keyboardavoidingview) being the most frequently used. For more complex or custom keyboard interactions, you might consider using [Keybaord Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), a library that offers advanced keyboard handling capabilities.

Reviewer: ```suggestion
React Native provides several keyboard-related APIs, with [Keyboard](https://reactnative.dev/docs/keyboard) and [KeyboardAvoidingView](https://reactnative.dev/docs/keyboardavoidingview) being the most frequently used. For more complex or custom keyboard interactions, you might consider using [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), a library that offers advanced keyboard handling capabilities.
```

## The basics

> ### Keybaord events

Reviewer: ```suggestion
### Keyboard events
```

> The Keybaord module allows you to listen for native events, react to them, and make changes to the keyboard, such as dismissing it.

Reviewer: ```suggestion
The Keyboard module allows you to listen for native events, react to them, and make changes to the keyboard, such as dismissing it.
```

To listen for keyboard events, use the `Keyboard.addListener` method. This method takes an event name and a callback function as arguments. The callback function will be called with the event data when the keyboard is shown or hidden.

> ```js

Reviewer 2: ````suggestion
```tsx
````

For concistency, let's use `tsx` as our primary code language for all snippets in this guide. We've recently migrated a lot of code examples to use TypeScript since that's one of the common things user ask for.

Also, (to be redundant in the review), it'd be nice to add code file names for all code blocks in this guide. For reference on how, please see [this section](https://github.com/expo/expo/blob/main/docs/README.md#add-code-block) in Docs' readme.

```js
import { KeyboardAvoidingView } from 'react-native';

export default function App() {
  return (
    <KeyboardAvoidingView behavior={Platform.OS === 'ios' ? 'padding' : undefined} style={{ flex: 1 }}>
      <TextInput placeholder="Type here..." />
    </KeyboardAvoidingView>;
  );
}
```

This example will automatically adjust the height of the `KeyboardAvoidingView` based on the keyboard height, so that the input is always visible.

> **info** When using BottomTabs for navigation in Android, you might notice that focusing on an input field causes the bottom tabs to be pushed above the keyboard. To address this issue add the `softwareKeyboardLayoutMode` property to your Android configuration in `app.json` and set it to `pan`.

```js app.json
"expo" {
  "android": {
    "softwareKeyboardLayoutMode": "pan"
  }
}
```

After adding this property, restart your app so that the changes take effect.

## Advanced keyboard handling

For more complex keyboard interactions, consider using the [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller) library. It offers additional functionality beyond the built-in React Native keyboard APIs.

> ### Installing keybaord controller

Reviewer: ```suggestion
### Installing keyboard controller
```

export default function AdvancedToolbar() {
  return (
    <>
      <KeyboardAwareScrollView bottomOffset={62} contentContainerStyle={styles.container}>
        <View>
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
        <View>
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
      </KeyboardAwareScrollView>
      <KeyboardToolbar />
    </>
  );
}
```

The example wraps the inputs with `KeyboardAwareScrollView` to prevent the keyboard from covering them. The `KeyboardToolbar` component displays navigation controls and a dismiss button. While it works without configuration, you can customize the toolbar content if needed.

> Keyboard Controller provides consistency across Android and iOS with minimal configuration, offering the native feel users expect.

Reviewer 2: ```suggestion
Keyboard Controller provides consistency across Android and iOS with minimal configuration, offering the native feel that users expect.
```

Let's move this line to the introduction of the Keyboard Controller section since it is telling about the advantage that this library provides.

## Additional resources

- [Keyboard Guide Source Code](https://github.com/betomoedano/keyboard-guide.git)

> - [Keybaord Controller Docs](https://kirillzyusko.github.io/react-native-keyboard-controller)

Reviewer: ```suggestion
- [Keyboard Controller Docs](https://kirillzyusko.github.io/react-native-keyboard-controller)
```

> - [Keyboard Guide Source Code](https://github.com/betomoedano/keyboard-guide.git)
> - [Keyboard Controller Docs](https://kirillzyusko.github.io/react-native-keyboard-controller)

Reviewer 2: ```suggestion
- [Keyboard guide source code](https://github.com/betomoedano/keyboard-guide)
- [Keyboard Controller documentation](https://kirillzyusko.github.io/react-native-keyboard-controller)
```

---
title: Keyboard handling
description: A guide for handling common keyboard interactions.
---

import { Terminal } from '~/ui/components/Snippet';

Keyboard handling is crucial for creating an excellent user experience in your Expo app. This guide covers common keyboard interactions and how to manage them effectively.

React Native provides several keyboard-related APIs, with [`Keyboard`](https://reactnative.dev/docs/keyboard) and [`KeyboardAvoidingView`](https://reactnative.dev/docs/keyboardavoidingview) being the most frequently used. For more complex or custom keyboard interactions, you might consider using [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), which is a library that offers advanced keyboard handling capabilities.

## The basics

> ### Keyboard events

Reviewer 3: I feel like the keyboard events are a less-frequent / more advanced use case than simple keyboard avoidance, so what if we put keyboard avoidance first and the events second?

The `Keyboard` module from React Native allows you to listen for native events, react to them, and make changes to the keyboard, such as dismissing it.

> To listen for keyboard events, use the `Keyboard.addListener` method. This method accepts an event name and a callback function as arguments. The callback function is called with the event data when the keyboard is shown or hidden.

Reviewer 3: Could we ground the keyboard events section in an example of why you'd want to use it? Not that this needs to affect the code example itself, I'm thinking something as simple as "you may want to know if the keyboard is visible or not visible because..."

Searching through some of my old code, one place where I used these events in conjunction with `Keyboard.dismiss` was to set a state variable to show a keyboard dismiss button in the header on a chat screen when they keyboard was visible on iOS, which doesn't have a built-in button for this. Then that button used `Keyboard.dismiss`.

Reviewer 3: or maybe another example would be using a `scrollable view ref to scroll the content to the bottom to show the latest message in a chat?

```tsx HomeScreen.tsx
import { useEffect } from 'react';
import { Keyboard } from 'react-native';

export default function HomeScreen() {
  useEffect(() => {
    const showSubscription = Keyboard.addListener('keyboardDidShow', handleKeyboardShow);

    return () => {
      showSubscription.remove();
    };
  }, []);

  const handleKeyboardShow = event => {
    // Handle keyboard show event here
  };
}
```

To dismiss the keyboard, use the `dismiss` method.

```tsx DismissKeyboardButton.tsx
import { Keyboard, Button } from 'react-native';

export default function DismissKeyboardButton() {
  return <Button title="Dismiss keyboard" onPress={Keyboard.dismiss} />;
}
```

### Keyboard avoiding view

The `KeyboardAvoidingView` is a component that automatically adjusts its height, position, or bottom padding based on the keyboard height to remain visible while the keyboard is displayed.

```tsx HomeScreen.tsx
import { KeyboardAvoidingView, TextInput } from 'react-native';

export default function HomeScreen() {
  return (

>     <KeyboardAvoidingView behavior={Platform.OS === 'ios' ? 'padding' : undefined} style={{ flex: 1 }}>

Reviewer 3: We should clarify what `undefined` means here ... I forget if it means that the keyboard avoiding view has no effect (which is often what you want with `adjustResize`) or it defaults to something. If what you want here changes when you adjust to `pan`, we should clarify that.

Broader goal here: we can improve a bit on the default RN `KeyboardAvoidingView` documentation to clarify what are typically good values to use for `behavior`. I would say this would be whatever properly keeps a single `TextInput` visible within a scrollable view on both iOS and Android with the default `adjustResize` soft input value is a good baseline.

      <TextInput placeholder="Type here..." />
    </KeyboardAvoidingView>;
  );
}
```

In the above example, the height of the `KeyboardAvoidingView` automatically adjusts based on the device's keyboard height, which ensures that the input is always visible.

When using a Bottom Tab navigator on Android, you might notice that focusing on an input field causes the bottom tabs to be pushed above the keyboard. To address this issue, add the `softwareKeyboardLayoutMode` property to your Android configuration in [app confg](/workflow/configuration/) and set it to `pan`.

```json app.json
"expo" {
  "android": {
    "softwareKeyboardLayoutMode": "pan"
  }
}
```

> After adding this property, restart your app so that the changes take effect.

Reviewer 3: This requires a development build and an actual rebuild of your app to take effect, right?

I was thinking we could also mention that `tabBarHideOnKeyboard` is an option that would still work with `adjustResize`, but might not look quite as smooth. https://reactnavigation.org/docs/bottom-tab-navigator/#tabbarhideonkeyboard

Author: It requires restarting the server; I need to make that clear. But this will work on Expo Go (no need for dev build)

Wow, I didn't know about `tabBarHideOnKeyboard`. We could mention it as an alternative, maybe.

## Advanced keyboard handling

> For more complex keyboard interactions, consider using the [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller) library. It offers additional functionality beyond the built-in React Native keyboard APIs, providing consistency across Android and iOS with minimal configuration and offering the native feel users expect.

Reviewer 3: ```suggestion
For more complex keyboard interactions, such as larger scrollable entry forms with several text input fields, consider using the [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller) library. It offers additional functionality beyond the built-in React Native keyboard APIs, providing consistency across Android and iOS with minimal configuration and offering the native feel users expect.
```
Suggestion to ground in an example

### Install keyboard controller

To get started with [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), install it in your Expo project:

<Terminal cmd={['$ npx expo install react-native-keyboard-controller']} />

Keyboard Controller uses native code, so you need to prebuild your app:

> <Terminal cmd={['$ npx expo prebuild --clean']} />

Reviewer 3: You would only need to explicitly run this if you were building locally with `npx expo run`. I would instead say that you will need to use a development build since this package is not in Expo Go. (then link to development build)

Reviewer 2: Adding to what Keith mentioned in the above comment, I think we can remove the prebuild command from here and instead:

- Add a subsection called "prerequisites" and mention that this guide assumes folks are creating development builds. Something like the following:

```md
The following steps described use a [development build](/develop/development-builds/introduction/) since this library is not included in Expo Go. See [create a development build](/develop/development-builds/create-a-build/) for more information.
```

Keyboard Controller library also requires `react-native-reanimated` to work properly. Follow the [installation instructions](/versions/latest/sdk/reanimated/#installation) to install it.

### Set up Keyboard Controller provider

To finalize the setup, add the `KeyboardProvider` to your app.

```tsx _layout.tsx
import { Stack } from 'expo-router';
import { KeyboardProvider } from 'react-native-keyboard-controller';

export default function RootLayout() {
  return (
    <KeyboardProvider>
      <Stack>
        <Stack.Screen name="basic" />
        <Stack.Screen name="advanced" />
        <Stack.Screen name="+not-found" />
      </Stack>
    </KeyboardProvider>
  );
}
```

> ### Usage

Reviewer 3: As I recall, you don't have to use it like this, right? Is this needed for simply accounting for keyboard space, or is it only needed when animating something else along with the keyboard, where frame-tracking is needed? 

If we include this, we should ground it in why you would want to do this instead of using `KeyboardAvoidingView` as opposed to how it is similar, because this is more complex than using `KeyboardAvoidingView`.

Reviewer 3: UPDATE: seeing where this fits in with the video demonstration, makes more sense. I'd still put this below the `KeyboardAwareScrollView` part, as that covers a very common `KeyboardAvoidingView` pain point that's very plug-and-play fixable. 

So yeah, I'd say change the heading to indicate what we're specifically trying to do (e.g., smoothly animate the view with the keyboard height) and introduce this section with the scenario described in the video so readers understand why you might want to do this.

Author: The provider is required! for these reasons: https://kirillzyusko.github.io/react-native-keyboard-controller/docs/recipes/platform-differences

Reviewer 3: yeah, I understand that the provider is required, I was just thinking that you don't always need to hook in with an animation worklet, that just wrapping with the provider opens the door to using something more basic like the `KeyboardAwareScrollView`.

Author: Oh, I understand what you're saying. Yes, it is possible to use `KeyboardAwareScrollView` (I just tested it, and it works fine with the chat screen). The downside is that if you use a `FlatList,` you'll get the `VirtualizedList` warning (see screenshot).

<img width="435" alt="image" src="https://github.com/user-attachments/assets/efbbe17b-632a-479c-8990-5149ad9c1e07">

We can mention this as a more accessible alternative solution, but for more extensive lists, a `ScrollView` won't be very performant. Using an animated style allows for extra customization and IMO; it feels more snappy.

  const { height } = useGradualAnimation();

  const fakeView = useAnimatedStyle(() => {
    return {
      height: Math.abs(height.value),
    };
  }, []);

  return (
    <View style={styles.container}>
      <FlatList
        data={messages}
        renderItem={({ item }) => <MessageItem message={item} />}
        keyExtractor={item => item.createdAt.toString()}
        contentContainerStyle={styles.listStyle}
      />
      <TextInput placeholder="Type a message..." style={styles.textInput} />
      <Animated.View style={fakeView} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,
  },
  listStyle: {
    padding: 16,
    gap: 16,
  },
  textInput: {
    width: "95%",
    height: 45,
    borderWidth: 1,
    borderRadius: 8,
    borderColor: "#d8d8d8",
    backgroundColor: "#fff",
    padding: 8,
    alignSelf: "center",
    marginBottom: 8,
  },
});
```

The above uses the keyboard's height to move the screen content above the keyboard. The view after the keyboard has an animated style that modifies its height based on the keyboard's height.

> ### Handling multiple inputs

Reviewer 3: This feels like the more basic/essential scenario than smoothly animating a toolbar input, maybe could go above that.

Author: You're right! Although this is also the last part of the video. Do you think having this part first in the guide and last in the video will be confusing? Maybe we can leave it as a bonus part?

Reviewer 3: I think it's OK if the video is in a different sequence. It's more like showing through a practical example than summarizing the different techniques.

---
title: Keyboard handling

> description: A guide for handling common keyboard interactions.

Reviewer 2: ```suggestion
description: A guide for handling common keyboard interactions on a device.
```

Let's add the context that this guide talks about keyboards on a real device.

---

> import { Terminal } from '~/ui/components/Snippet';

Reviewer 2: ```suggestion
import { GithubIcon } from '@expo/styleguide-icons/custom/GithubIcon';

import { BoxLink } from '~/ui/components/BoxLink';
import { Terminal } from '~/ui/components/Snippet';
import { CODE } from '~/ui/components/Text';
```

Let's add these import statements so that we can add BoxLinks in Additiona resources section.

        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
        <View>
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
      </KeyboardAwareScrollView>
      <KeyboardToolbar />
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    gap: 16,
    padding: 16,
  },
  listStyle: {
    padding: 16,
    gap: 16,
  },
  textInput: {
    width: 'auto',
    flexGrow: 1,
    flexShrink: 1,
    height: 45,
    borderWidth: 1,
    borderRadius: 8,
    borderColor: '#d8d8d8',
    backgroundColor: '#fff',
    padding: 8,
    marginBottom: 8,
  },
});
```

The above example wraps the inputs with `KeyboardAwareScrollView` to prevent the keyboard from covering them. The `KeyboardToolbar` component displays navigation controls and a dismiss button. While it works without configuration, you can customize the toolbar content if needed.

## Additional resources

> - [Example project based on this guide](https://github.com/betomoedano/keyboard-guide)
> - [`react-native-keyboard-controller` documentation](https://kirillzyusko.github.io/react-native-keyboard-controller)

Reviewer 2: ```suggestion
<BoxLink
  title="Example"
  description="See the source code for the example project on GitHub."
  Icon={GithubIcon}
  href="https://github.com/betomoedano/keyboard-guide"
/>

<BoxLink
  title={<CODE>react-native-keyboard-controller</CODE>}
  description="GitHub repository for the Keyboard Controller library."
  Icon={GithubIcon}
  href="https://github.com/kirillzyusko/react-native-keyboard-controller"
/>
```

Let's convert these links into BoxLinks

### Prerequisites

The following steps are described using a [development build](/develop/development-builds/introduction/) since this library is not included in Expo Go. See [Create a development build](/develop/development-builds/create-a-build/) for more information.

[Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller) requires `react-native-reanimated` to work correctly. To install it, follow the [installation instructions](/versions/latest/sdk/reanimated/#installation).

### Install keyboard controller

To get started with [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), install it in your Expo project:

<Terminal cmd={['$ npx expo install react-native-keyboard-controller']} />

Then prebuild your app:

<Terminal cmd={['$ npx expo prebuild --clean']} />

> ### Set up Keyboard Controller provider

Reviewer 4: ```suggestion
### Set up keyboard controller provider
```

Nit, ref: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#headings

To finalize the setup, add the `KeyboardProvider` to your app.

```tsx _layout.tsx
import { Stack } from 'expo-router';
import { KeyboardProvider } from 'react-native-keyboard-controller';

export default function RootLayout() {
  return (
    <KeyboardProvider>
      <Stack>
        <Stack.Screen name="home" />
        <Stack.Screen name="chat" />
      </Stack>
    </KeyboardProvider>
  );
}
```

> ### Animating Views in Sync with Keyboard Height

Reviewer 4: ```suggestion
### Animating views in sync with keyboard height
```

The `KeyboardAvoidingView` component is excellent for prototyping but requires platform-specific configuration and is not very customizable. Here's how to achieve the same functionality using [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller).

Let's use the `useKeyboardHandler` hook from [Keyboard Controller](https://kirillzyusko.github.io/react-native-keyboard-controller), which provides access to keyboard lifecycle events and allows us to determine when the keyboard starts animating and its position in every frame of the animation.
The example below is a hook that leverages the `useKeyboardHandler` hook to access the height of the keyboard at each frame, using `useSharedValue` from reanimated returns the height.

```tsx ChatScreen.tsx
import { useKeyboardHandler } from 'react-native-keyboard-controller';
import Animated, { useAnimatedStyle, useSharedValue } from 'react-native-reanimated';

const useGradualAnimation = () => {
  const height = useSharedValue(0);

  useKeyboardHandler(
    {

>       onMove: e => {
>         'worklet';
>         height.value = Math.max(e.height, 0);
>       },

Reviewer 4: ```suggestion
      onMove: (event) => {
        'worklet';
        height.value = Math.max(event.height, 0);
      },
```

Naming nit.

```tsx ChatScreen.tsx
import { StyleSheet, Platform, FlatList, View, StatusBar, TextInput } from "react-native";
import Animated, { useAnimatedStyle, useSharedValue } from "react-native-reanimated";
import { useKeyboardHandler } from "react-native-keyboard-controller";

import MessageItem from "@/components/MessageItem";
import { messages } from "@/messages";

const useGradualAnimation = () => {...};

export default function ChatScreen() {
  const { height } = useGradualAnimation();

  const fakeView = useAnimatedStyle(() => {
    return {
      height: Math.abs(height.value),
    };
  }, []);

  return (
    <View style={styles.container}>
      <FlatList
        data={messages}
        renderItem={({ item }) => <MessageItem message={item} />}
        keyExtractor={item => item.createdAt.toString()}
        contentContainerStyle={styles.listStyle}
      />
      <TextInput placeholder="Type a message..." style={styles.textInput} />
      <Animated.View style={fakeView} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,

>     paddingTop: Platform.OS === "android" ? StatusBar.currentHeight : 0,

Reviewer 2: ```suggestion
    paddingTop: Platform.OS === 'android' ? StatusBar.currentHeight : 0,
```

Same as previous

  },
  listStyle: {
    padding: 16,
    gap: 16,
  },
  textInput: {

>     width: "95%",
>     height: 45,
>     borderWidth: 1,
>     borderRadius: 8,
>     borderColor: "#d8d8d8",
>     backgroundColor: "#fff",
>     padding: 8,
>     alignSelf: "center",
>     marginBottom: 8,

Reviewer 2: ```suggestion
    width: '95%',
    height: 45,
    borderWidth: 1,
    borderRadius: 8,
    borderColor: '#d8d8d8',
    backgroundColor: '#fff',
    padding: 8,
    alignSelf: 'center',
    marginBottom: 8,
```

Same as previous

        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
        <View>
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
          <TextInput placeholder="Type a message..." style={styles.textInput} />
        </View>
        <TextInput placeholder="Type a message..." style={styles.textInput} />
      </KeyboardAwareScrollView>
      <KeyboardToolbar />
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    gap: 16,
    padding: 16,
  },
  listStyle: {
    padding: 16,
    gap: 16,
  },
  textInput: {
    width: 'auto',
    flexGrow: 1,
    flexShrink: 1,
    height: 45,
    borderWidth: 1,
    borderRadius: 8,
    borderColor: '#d8d8d8',
    backgroundColor: '#fff',
    padding: 8,
    marginBottom: 8,
  },
});
```

The above example wraps the inputs with `KeyboardAwareScrollView` to prevent the keyboard from covering them. The `KeyboardToolbar` component displays navigation controls and a dismiss button. While it works without configuration, you can customize the toolbar content if needed.

## Additional resources

> - [Keyboard guide source code](https://github.com/betomoedano/keyboard-guide)
> - [Keyboard Controller documentation](https://kirillzyusko.github.io/react-native-keyboard-controller)

Reviewer 4: Presentation nit, we can convert those inline links to `BoxLink`s.

```suggestion
- [Sample project based on this guide](https://github.com/betomoedano/keyboard-guide)
- [`react-native-keyboard-controller` documentation](https://kirillzyusko.github.io/react-native-keyboard-controller)
```

Also, small suggestion to the wording, and in case of docs link, it seems more clear to use the package name, since  "Keyboard Controller" sounds quite generic.

The `Keyboard` module from React Native allows you to listen for native events, react to them, and make changes to the keyboard, such as dismissing it.

To listen for keyboard events, use the `Keyboard.addListener` method. This method accepts an event name and a callback function as arguments. When the keyboard is shown or hidden, the callback function is called with the event data.

The following example illustrates a use case for adding a keyboard listener. The state variable `isKeyboardVisible` is toggled each time the keyboard shows or hides. Based on this variable, a button allows the user to dismiss the keyboard only if the keyboard is active. Also, notice that the button uses the `Keyboard.dismiss` method.

```tsx HomeScreen.tsx
import { useEffect, useState } from 'react';
import { Keyboard, View, Button, TextInput } from 'react-native';

export default function HomeScreen() {
  const [isKeyboardVisible, setIsKeyboardVisible] = useState(false);

  useEffect(() => {
    const showSubscription = Keyboard.addListener('keyboardDidShow', handleKeyboardShow);
    const hideSubscription = Keyboard.addListener('keyboardDidHide', handleKeyboardHide);

    return () => {

>       showSubscription.remove();

Reviewer 5: Should we also remove the hideSubscription?
```suggestion
      showSubscription.remove();
      hideSubscription.remove();
```

    };
  }, []);

  const handleKeyboardShow = event => {
    setIsKeyboardVisible(true);
  };

  const handleKeyboardHide = event => {
    setIsKeyboardVisible(false);

>   };

Reviewer 5: Correct me if I'm wrong, but shouldn't these two either be defined inside the useEffect function or else added to the deps array?
